libkvmi integration

adlazar opened the pull request:
It builds fine, but I didn't test it yet.
The comments are welcomed.

adlazar committed:
libvmi/drivers/kvm: add libkvmi

Signed-off-by: Mihai Donțu <mdontu@bitdefender.com>
Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>
Signed-off-by: Nicușor Cîțu <ncitu@bitdefender.com>
Signed-off-by: Mircea Cîrjaliu <mcirjaliu@bitdefender.com>
Signed-off-by: Marian Rotariu <mrotariu@bitdefender.com>

adlazar committed:
libvmi/driver/kvm: add the KVM headers used by libkvmi

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

adlazar committed:
libvmi/driver/kvm: integrate libkvmi

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Wenzel commented on libvmi/driver/kvm/kvm.c:
```diff
@@ -1436,6 +1623,15 @@ kvm_init_vmi(
+        dbprint(VMI_DEBUG_KVM, "--KVMI failed\n");
+        return VMI_FAILURE;
+    }
+    dbprint(VMI_DEBUG_KVM, "--KVMI connected \o/\n");
```
I understand your excitation, i share it, but i think you can remove the emoji \o/ here ;)

adlazar committed:
remove improperly escaped emoji

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

adlazar committed:
libvmi/driver/kvm: cleanup some libkvmi related comments

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Wenzel commented on configure.ac:
```diff
@@ -87,6 +87,13 @@ AC_ARG_ENABLE([kvm],
       [enable_kvm=yes])
 AM_CONDITIONAL([WITH_KVM], [test x"$enable_kvm" = xyes])
 
+AC_ARG_ENABLE([libkvmi],
```
I don't think that libkvmi should be enabled or disabled.

To me it **replaces** the current KVM driver of Libvmi, so if the user configured Libvmi to have KVM support, just check that libkvmi is here, or throw an error.

And in the driver there is no need for `HAVE_LIBKVMI` in the same way.

I would like to have @tklengyel opinion too.

tklengyel replied:
I would be all for replacing the existing KVM driver with this, but there are features built on top of the existing KVM driver that would need to be removed in that case, like SHM. IMHO that feature was always experimental and not well maintained so removing it should be fine.

namidairo commented:
This does not appear to be hooked up via the driver interface yet, so it's still gonna call all the old code.

adlazar commented:
It's more like a demo on how to integrate the kvmi library. :) Maybe I should change the title.
The new functions are called from `kvm_init_vmi`, `kvmi_destroy` and `kvm_get_vcpureg`.

Soft committed:
Fix MSR indexes to correspond kernel's msr-index.h

Soft committed:
Add missing break statement and MSR_STAR handling to get_kvmi_registers.

Soft committed:
Fix kvmi_get_registers return value handling

Soft committed:
Partial implementation of kvm_get_vcpuregs.

This will enable querying all the registers in one step.

Soft committed:
Revert previous MSR index fix as it conflicted with Xen driver.

Implement a new KVM driver specific MSR translation.

This reverts commit f75a73c31fe06c8c29f41fb31ce32c93babb92a3.

adlazar committed:
Merge pull request #1 from Soft/kvmi

Fix MSR indexes to correspond kernel's msr-index.h

Soft committed:
More correct typing for translate_msr_index

Soft committed:
Add memory access support for kvmi driver

Soft committed:
Added support for setting register values.

This is still quite experimental. To be more specific, I think kvm_set_vcpureg
is likely to be broken. The reason for this is that the kernel only supports
setting all the registers at once and for us to be able to set only a single
register we need to retrieve the old values first. Of course, both of these
operations need to be done without the vcpu continuing execution while the
operations are happening. Maybe we should explicitly pause the vcpu before the
operations.

adlazar committed:
Merge pull request #2 from Soft/kvmi

Memory & Register Access

Wenzel commented:
What's the status of this PR ?
What concrete features does it bring ATM ?
Can we merge it already, or are you still polishing the driver ?

adlazar commented:
@Wenzel, my intention with this PR was to initiate the `libkvmi` integration (init, uninit, domain callbacks ...). It wasn't meant as a complete replacement of the current introspection methods. I see that by not using a proper branch name (i.e. kvmi-begin-libkvmi-integration) and merging @Soft's work it makes this PR look like a replacement. Also, dropping --enable-libkvmi is not an issue, just that I wasn't courageous enough to replace the old introspection methods from this driver :)

I could close this PR if you want to merge it only when is done.

Wenzel commented:
> @Wenzel, my intention with this PR was to initiate the libkvmi integration (init, uninit, domain callbacks ...). It wasn't meant as a complete replacement of the current introspection methods. I see that by not using a proper branch name (i.e. kvmi-begin-libkvmi-integration) and merging @Soft's work it makes this PR look like a replacement.

I would rather merge what we already have, so we can build tests for libvmi, than continuing adding features on this PR.

This will bring me more visibility, because ATM i don't know what feature you are implementing next @Soft  :)
Let's keep the PR short.
one feature, one PR, and a unit test to validate.

What do you guys think ?

Soft committed:
Fix broken header.

Apparently I had broken the kvm.h header when picking chunks for the previous
commit. Sorry for that.

Soft committed:
Update .gitignore to exclude unnecessary cruft.

adlazar committed:
Merge pull request #3 from Soft/kvmi

Fix broken header

adlazar committed:
Merge pull request #4 from Soft/gitignore-update

Update .gitignore to exclude unnecessary cruft.

Soft committed:
Fix double free in kvmi_uninit.

kvmi_uninit was called twice leading to a double free bug if unix socket
connection errored. This fixes the problem by manually setting the kvmi pointer
to NULL as kvmi_uninit checks for that.

Soft committed:
Allow passing socket path to init_kvmi.

This replaces the hard coded /var/run/testing.sock path with a configurable option.

adlazar committed:
Merge pull request #5 from Soft/kvmi

Fix double free in kvmi_uninit.

adlazar committed:
libvmi/driver/kvm: drop --enable-libkvmi

Revert most of the changes from commit f392f4aa6623
('libvmi/driver/kvm: integrate libkvmi'),
but let HAVE_LIBKVMI to be removed later.

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Soft committed:
Comment out Qemu version check and add process-list example with configurable socket path.

Soft committed:
Only kvmi based access in kvm driver.

This commit removes all the other access methods from the kvm driver and only
leaves the new kvmi based one. Now `--enable-kvm` implies kvmi.

Soft committed:
Implement kvm_put_memory using kvmi

Wenzel commented:
ATM i can't test this PR because `libvmi` won't initialize:
![screenshot from 2018-08-06 13-36-31](https://user-images.githubusercontent.com/964610/43712129-d00abe38-997d-11e8-8ea3-d17f3b39c579.png)

This check happens [here](https://github.com/libvmi/libvmi/blob/master/libvmi/driver/kvm/kvm.c#L1482), to verify that if the current `QEMU` version is greater than `2.8.0` and we don't have detected `libvmi_request` headers installed by `QEMU` (`/usr/include/qemu/libvmi_request.h`) we exit, because we are supposed to use `libvmi_resquest.h` from `2.8.0` and more for memory access.

![screenshot from 2018-08-06 13-43-08](https://user-images.githubusercontent.com/964610/43712414-c73419d4-997e-11e8-8e26-40a18edefe8b.png)

That's why i want to remove all of this code for legacy memory access methods, it's very confusing :)
@tklengyel can give his opinion maybe ?

In the meantime, @Soft created a dedicated branch [kvmi-fixes](https://github.com/Soft/libvmi/tree/kvmi-fixes) which removes this code and allows us to test the driver.

Soft commented:
Just for the record, there is also another [branch](https://github.com/Soft/libvmi/tree/kvmi-only) that removes all the other KVM introspection methods from the KVM driver leaving only kvmi specific introspection code.

adlazar commented:
Let's merge https://github.com/Soft/libvmi/tree/kvmi-only. @tklengyel already said (https://github.com/KVM-VMI/libvmi/pull/2/files/f392f4aa6623fb64a5853eef25718427ad5192b3#r197895268) that he is for the replacement of the old KVM driver. Right?

Wenzel commented:
@adlazar i approve, it's just that @tklengyel's message was confusing, but i think he meant to replace it all.
@Soft can you make a PR to @adlazar 's repo with your `kvmi-only` branch ?
Note: if i remove the QEMU version check, `libkvmi`'s connection fails:
![screenshot from 2018-08-06 16-11-54](https://user-images.githubusercontent.com/964610/43718546-a8cad8ce-9993-11e8-9502-432c42384b14.png)

Wenzel commented:
This PR is waiting on https://github.com/adlazar/libvmi/pull/8 to be merged.

Soft committed:
Remove invalid reference to --enable-libkvmi

adlazar committed:
Merge branch 'kvmi' into kvmi-only

adlazar committed:
Merge pull request #8 from Soft/kvmi-only

Remove other KVM access methods

Wenzel commented:
Thanks @Soft and @adlazar for merging the `kvmi-only` branch, i'm rebuilding my environment from scratch to test the changes.

Wenzel commented:
The example is working !
Good job guys :+1: 
![confirm](https://user-images.githubusercontent.com/964610/44460017-398e8b80-a614-11e8-8d5d-310491d54ee5.png)

namidairo commented:
Just skimming through this and noticed that you might have re-introduced the bug where the last part of the memory becomes inaccessible, because of the assumption max memory from libvirt translates to last addressable memory.

I guess that'd break things on a vm using near it's total allocated memory?

Wenzel commented:
@namidairo do you mean this bug was already fixed in libvmi `master` ?
Can you give me some reference ? (commit, code, ...) because i can't remember it.

namidairo commented:
https://github.com/libvmi/libvmi/commit/2975d22717b0f51f4e18f9267191aeef691ef123#diff-4ec4791b3c1d65a6821f6146d30de5de

All the QMP code got removed in this PR, so I'm not sure if we'd put it back just for this or just expand on the qemu patch since kvmi has one already.
